### ISSUE TITLE
Mechanism error while trying to add new card

### ISSUE THREAD

**Issue by StipT** [PeculiarVentures/fortify#312]
### I encountered some kind of mechanism error while adding new card in card.json
*Encountered on Windows10 Pro x64
**Certificate works perfectly when accessed from Windows CryptoAPI provider.
```
    {
      "atr": "3bff1300008131fe450031b9640444ecc17394018082900012",
      "name": "Croatian eID",
      "driver": "941BB1B17DFC1FF2267920A8535D2D809F0C5A0B"
    }
```
```
    {
      "id": "941BB1B17DFC1FF2267920A8535D2D809F0C5A0B",
      "name": "Croatian eID",
      "file": {
        "windows": "%PROGRAMFILES/AKD/eID Middleware/pkcs11/AkdEidPkcs11_64.dll",
        "osx": "/usr/local/lib/pkcs11/libEidPkcs11.dylib"
      }
```
**LOG:**

```
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: reader:'Generic Smart Card Reader Interface 0' name:'Croatian eID' atr:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: Loading PKCS#11 library from C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll","level":"info"}
{"message":"Provider:Token:Insert: Looking for Generic Smart Card Reader Interface 0 into 2 slot(s)","level":"info"}
{"message":"Error: CKR_MECHANISM_INVALID:112\n    at Error (native) PKCS11::C_GetMechanismInfo:279\n    at Mechanism.getInfo (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:86:31)\n    at new Mechanism (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:83:14)\n    at MechanismCollection.items (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:101:16)\n    at Object.getProviderInfo (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\utils.js:76:38)\n    at Pkcs11Crypto.open (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\crypto.js:59:27)\n    at new Crypto (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\crypto.js:42:14)\n    at new Pkcs11Crypto (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:1892:9)\n    at C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:2221:40\n    at Array.forEach (<anonymous>)\n    at LocalProvider.onTokenInsert (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:2219:29)","level":"error"}
{"message":"Error: CKR_MECHANISM_INVALID:112\n    at Error (native) PKCS11::C_GetMechanismInfo:279\n    at Mechanism.getInfo (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:86:31)\n    at new Mechanism (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:83:14)\n    at MechanismCollection.items (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:101:16)\n    at Object.getProviderInfo (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\utils.js:76:38)\n    at Pkcs11Crypto.open (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\crypto.js:59:27)\n    at new Crypto (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\crypto.js:42:14)\n    at new Pkcs11Crypto (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:1892:9)\n    at C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:2221:40\n    at Array.forEach (<anonymous>)\n    at LocalProvider.onTokenInsert (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:2219:29)","level":"error"}
```

**Comment by rmhrisk** [comment]
@microshine

**Comment by microshine** [comment]
Please get a list of mechanisms. Looks like your PKCS#11 supports vendor mechanisms

### Install graphene console application
```
> npm i graphene-cli -g
```

### Run console application
```
> graphene
```

### Initialize PKCS#11 module
```
> module load --lib /usr/local/lib/pkcs11/libEidPkcs11.dylib --name "Croatian eID"
```

### List all mechanisms
```
> slot algs --slot 0
```

#### Example output for SoftHSM
```
| Algorithm name            | h | s | v | e | d | w | u | g | D |
|---------------------------|---|---|---|---|---|---|---|---|---|
| MD5                       | x |   |   |   |   |   |   |   |   |
| SHA_1                     | x |   |   |   |   |   |   |   |   |
| SHA224                    | x |   |   |   |   |   |   |   |   |
| SHA256                    | x |   |   |   |   |   |   |   |   |
| SHA384                    | x |   |   |   |   |   |   |   |   |
| SHA512                    | x |   |   |   |   |   |   |   |   |
| MD5_HMAC                  |   | x | x |   |   |   |   |   |   |
| SHA_1_HMAC                |   | x | x |   |   |   |   |   |   |
...
```

**Comment by StipT** [comment]
```
> module load --lib C:\Program Files\AKD\eID Middleware\pkcs11\AkdEidPkcs11_64.dll --name "Croatian eID"

Error Win32 error 126
    at Error (native) PKCS11::Load:136

Help:
  loads a specified PKCS#11 module

  --lib          -l     Path to library
  --name         -n     Name of module

  Commands:
    ?              Prints help information
```
Any idea what am I doing wrong here? Also, I've seen on [recent post](https://github.com/PeculiarVentures/fortify/issues/292) similar issue with the same card so I guess it's the same problem.

**Comment by microshine** [comment]
Do you use NodeJs x85 or x64 architecture?

**Comment by StipT** [comment]
x64

**Comment by microshine** [comment]
It can be useful https://stackoverflow.com/a/14362289/2934842

I'll try to install the same middleware and test it locally (if it's possible without token)

**Comment by microshine** [comment]
I installed AKD eID to my Windows machine and ran this script.
```js
var graphene = require("graphene-pk11")

function main() {
  const lib = "C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll";

  const reader = graphene.Module.load(lib);

  reader.initialize();

  const slots = reader.getSlots();
  console.log(slots.length);
}

try {
  main();
  process.exit(0);
} catch (e) {
  console.error(e);
  process.exit(1);
}
```

For the first start, I got an error with code 193.
```log
Error: Win32 error 193
    at Error (native) PKCS11::Load:136
```
After that script works fine.

Please check out the path to your PKCS#11 library. It throws `Win32 error 126` if the file does not exist.

Try
```
module load --lib "C:\Program Files\AKD\eID Middleware\pkcs11\AkdEidPkcs11_64.dll" --name "Croatian eID"
```

**Comment by StipT** [comment]
I moved AkdEdePkcs11_64.dll to \System32 where all dependencies were.
![dll_depends](https://i.imgur.com/AKEvZEL.jpg)
... and it worked.

```
C:\WINDOWS\system32>graphene
> module load --lib C:\Windows\System32\AkdEidPkcs11_64.dll --name "Croatian eID"

Module info
==============================

  Library: C:\Windows\System32\AkdEidPkcs11_64.dll
  Name: Croatian eID
  Cryptoki version: 2.20

> slot algs --slot 0

| Algorithm name            | h | s | v | e | d | w | u | g | D |
|---------------------------|---|---|---|---|---|---|---|---|---|
| RSA_PKCS                  |   | x |   |   | x |   |   |   |   |

Error CKR_MECHANISM_INVALID:112
    at Error (native) PKCS11::C_GetMechanismInfo:279
```

**Comment by microshine** [comment]
It's very strange. PKCS#11 module throws an exception on getting information about its own mechanism

Also, it implements Cryptoki version 2.2, but `pcks11js` supports version 2.3

Please create a simple nodejs project

```
mkdir fortify-issue-312
cd fortify-issue-312
npm init --y
npm install graphene-pk11
```

Add file index.js
```js
var graphene = require("graphene-pk11")

function main() {
  const lib = "C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll";

  const reader = graphene.Module.load(lib);

  reader.initialize();

  const slot = reader.getSlots(0);
  console.log(slot.getMechanisms().items_);
}

try {
  main();
  process.exit(0);
} catch (e) {
  console.error(e);
  process.exit(1);
}
```
Run app
```
node index.js
```

It'll print mechanisms identifiers without getting information about each mechanism

![image](https://user-images.githubusercontent.com/7333669/80602820-4645c280-8a38-11ea-82bf-9351fb027efe.png)

**Comment by StipT** [comment]
```
[
    1,   3, 528, 544,
  592, 608, 624      
]
```

**Comment by microshine** [comment]
Your mechanisms are:
- `CKM_RSA_PKCS`
- `CKM_RSA_X_509` (throws CKR_MECHANISM_INVALID on C_GetMechanismInfo)
- `CKM_MD5`
- `CKM_SHA_1`
- `CKM_SHA256`
- `CKM_SHA384`
- `CKM_SHA512`

**Comment by StipT** [comment]
Is there anything that can be done for this smartcard to be a part of your supported roster?

**Comment by microshine** [comment]
I created an issue to fix this problem. I'll fix it and share a new build with you to test it

**Comment by microshine** [comment]
Please try a new build
https://drive.google.com/open?id=1yic3CrN2FT9hsbMSJHf6tSiCvtUyaO8n

**Comment by StipT** [comment]
At first glance everything looked good, just like when accessed via Windows CryptoAPI. But when i tried to create a signature this happened:
```
CryptoServerError: CKR_MECHANISM_INVALID
    at t.<anonymous> (https://unpkg.com/@webcrypto-local/client@1.0.14/build/webcrypto-socket.min.js:15:82579)
    at tryCatch (https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/7.7.0/polyfill.min.js:1:93250)
    at Generator.invoke [as _invoke] (https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/7.7.0/polyfill.min.js:1:93024)
    at Generator.t.<computed> [as next] (https://cdnjs.cloudflare.com/ajax/libs/babel-polyfill/7.7.0/polyfill.min.js:1:93782)
    at a (https://unpkg.com/@webcrypto-local/client@1.0.14/build/webcrypto-socket.min.js:1:321)
    at s (https://unpkg.com/@webcrypto-local/client@1.0.14/build/webcrypto-socket.min.js:1:524)
```

**Comment by microshine** [comment]
please get log from Fortify for that error. You can enable logging using Fortify icon in tray

**Comment by StipT** [comment]
```
{"message":"Application started at Wed Apr 29 2020 21:07:51 GMT+0200 (srednjoeuropsko ljetno vrijeme)","level":"info"}
{"message":"OS win32 x64 ","level":"info"}
{"message":"Fortify v1.0.22","level":"info"}
{"message":"System locale is 'hr'","level":"info"}
{"message":"Locale: Set language to 'en'","level":"info"}
{"message":"Fortify: Create window index","level":"info"}
{"message":"Update: Check for new update","level":"info"}
{"message":"Update: New version wasn't found","level":"info"}
{"message":"SSL certificate is loaded","level":"info"}
{"message":"Comparing current version of card.json file with remote","level":"info"}
{"message":"card.json has the latest version","level":"info"}
{"message":"Server: Started at 127.0.0.1:31337","level":"info"}
{"message":"Provider: Add crypto 'Windows CryptoAPI' 498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\Fortify\\pvpkcs11.dll' 'C:\\Program Files\\Fortify\\pvpkcs11.dll'","level":"info"}
{"message":"Provider: Add crypto 'NSS Certificate DB' a7370eae6951997646c5bfedf8f3df0d8b8b698d","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'softokn3.dll' 'softokn3.dll'","level":"info"}
{"message":"Provider:Opened","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:true","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/export","level":"info"}
{"message":"Provider:Token:Insert: reader:'Generic Smart Card Reader Interface 0' name:'Croatian eID' atr:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: Loading PKCS#11 library from C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/export","level":"info"}
{"message":"Provider:Token:Insert: Looking for Generic Smart Card Reader Interface 0 into 2 slot(s)","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider:Token Amount of tokens was changed (+2/-0)","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: New session connect https://digital-signature-app.herokuapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://digital-signature-app.herokuapp.com id:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc authorized:true","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc server/isLoggedIn","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider/action/info","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider/action/getCrypto","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:498cecf6fc0e9d62ba7fcf8e7d23f819ec02cec4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider/action/getCrypto","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/login","level":"info"}
{"message":"Fortify: Create window p11-pin","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider/action/getCrypto","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/getItem","level":"info"}
{"message":"Server: session:89a8971de41bacb884160f87d5ce0f9ccc6ba5023658a710e73c091a4218afcc provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/subtle/sign","level":"info"}
{"message":"Error: CKR_MECHANISM_INVALID:112\n    at Error (native) PKCS11::C_GetMechanismInfo:279\n    at Mechanism.getInfo (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:86:31)\n    at new Mechanism (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:31:14)\n    at MechanismCollection.items (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\graphene-pk11\\build\\mech.js:101:16)\n    at Function.getAlgorithm (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\mechs\\rsa\\crypto.js:114:42)\n    at C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\mechs\\rsa\\rsa-ssa.js:23:53\n    at new Promise (<anonymous>)\n    at RsaSsaProvider.<anonymous> (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\node-webcrypto-p11\\build\\mechs\\rsa\\rsa-ssa.js:20:20)\n    at Generator.next (<anonymous>)\n    at C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\tslib\\tslib.js:113:75\n    at new Promise (<anonymous>)","level":"error"}
{"message":"Server: Close session  origin:https://digital-signature-app.herokuapp.com (code: 1001)","level":"info"}
```

**Comment by microshine** [comment]
@StipT  I found a problem. It the same `mechanisms.items(i)` usage. I'll fix it tomorrow and share a new build

Thank you for your help

**Comment by microshine** [comment]
@StipT Please try a new build [fortify-win32-x64-v1_0_22.msi](https://drive.google.com/open?id=1yic3CrN2FT9hsbMSJHf6tSiCvtUyaO8n)

**Comment by StipT** [comment]
**Both tokens(Identity & Signing) from my smartcard work. But something else came up.**

### 1.
If fortify starts,  on autostart or manually, while the smartcard is already inserted -> application freezes. At that point, I can't do anything but exit the application and start it again without the card being inserted. Only after the application boots, I can insert the card.

**LOG:**
```
{"message":"Application started at Mon May 04 2020 19:12:35 GMT+0200 (srednjoeuropsko ljetno vrijeme)","level":"info"}
{"message":"OS win32 x64 ","level":"info"}
{"message":"Fortify v1.0.22","level":"info"}
{"message":"System locale is 'hr'","level":"info"}
{"message":"Locale: Set language to 'en'","level":"info"}
{"message":"Fortify: Create window index","level":"info"}
{"message":"Update: Check for new update","level":"info"}
{"message":"Update: New version wasn't found","level":"info"}
{"message":"SSL certificate is loaded","level":"info"}
{"message":"Comparing current version of card.json file with remote","level":"info"}
{"message":"card.json has the latest version","level":"info"}
{"message":"Server: Started at 127.0.0.1:31337","level":"info"}
{"message":"Provider: Add crypto 'Windows CryptoAPI' 59b38d98b9bf1c8667a8caef2d979b5e692de148","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\Fortify\\pvpkcs11.dll' 'C:\\Program Files\\Fortify\\pvpkcs11.dll'","level":"info"}
{"message":"Provider: Add crypto 'NSS Certificate DB' a7370eae6951997646c5bfedf8f3df0d8b8b698d","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'softokn3.dll' 'softokn3.dll'","level":"info"}
{"message":"Provider:Opened","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: reader:'Generic Smart Card Reader Interface 0' name:'Croatian eID' atr:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: Loading PKCS#11 library from C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll","level":"info"}
{"message":"Provider:Token:Insert: Looking for Generic Smart Card Reader Interface 0 into 2 slot(s)","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider:Token Amount of tokens was changed (+2/-0)","level":"info"}
{"message":"TypeError: Cannot assign to read only property 'extractable' of object '#<CryptoKey>'\n    at FileStorage.ecKeyToBase64 (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:233:29)\n    at FileStorage.saveIdentities (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:165:61)\n    at processTicksAndRejections (internal/process/task_queues.js:86:5)\n    at async FileStorage.getIdentity (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:193:13)\n    at async Server.getRandomBundle (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:604:26)\n    at async C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:445:40","level":"error"}
```
### 2.
Every time i click "reload" on Fortify Tools page, token disappears. Provider is still listed as "Croatian eID", but token does not show up. Also more often than not, the same thing happens when I refresh tools page or any other page that uses fortify,

### 3.
Not sure if this is related, but now I have to give access to a page with 2key permission code every single time I go to or refresh page that uses fortify.

**LOG**:
```
{"message":"Application started at Mon May 04 2020 19:27:38 GMT+0200 (srednjoeuropsko ljetno vrijeme)","level":"info"}
{"message":"OS win32 x64 ","level":"info"}
{"message":"Fortify v1.0.22","level":"info"}
{"message":"System locale is 'hr'","level":"info"}
{"message":"Locale: Set language to 'en'","level":"info"}
{"message":"Fortify: Create window index","level":"info"}
{"message":"Update: Check for new update","level":"info"}
{"message":"Update: New version wasn't found","level":"info"}
{"message":"SSL certificate is loaded","level":"info"}
{"message":"Comparing current version of card.json file with remote","level":"info"}
{"message":"card.json has the latest version","level":"info"}
{"message":"Server: Started at 127.0.0.1:31337","level":"info"}
{"message":"Provider: Add crypto 'Windows CryptoAPI' a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\Fortify\\pvpkcs11.dll' 'C:\\Program Files\\Fortify\\pvpkcs11.dll'","level":"info"}
{"message":"Provider: Add crypto 'NSS Certificate DB' a7370eae6951997646c5bfedf8f3df0d8b8b698d","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'softokn3.dll' 'softokn3.dll'","level":"info"}
{"message":"Provider:Opened","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"PCSCWatcher: New reader detected Generic Smart Card Reader Interface 0","level":"info"}
{"message":"TypeError: Cannot assign to read only property 'extractable' of object '#<CryptoKey>'\n    at FileStorage.ecKeyToBase64 (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:233:29)\n    at FileStorage.saveIdentities (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:165:61)\n    at processTicksAndRejections (internal/process/task_queues.js:86:5)\n    at async FileStorage.getIdentity (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:193:13)\n    at async Server.getRandomBundle (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:604:26)\n    at async C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:445:40","level":"error"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:false","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/login","level":"info"}
{"message":"Fortify: Create window key-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/certificateStorage/export","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"PCSCWatcher:Insert reader:'Generic Smart Card Reader Interface 0' ATR:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: reader:'Generic Smart Card Reader Interface 0' name:'Croatian eID' atr:3bff1300008131fe450031b9640444ecc17394018082900012","level":"info"}
{"message":"Provider:Token:Insert: Loading PKCS#11 library from C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll","level":"info"}
{"message":"Provider:Token:Insert: Looking for Generic Smart Card Reader Interface 0 into 2 slot(s)","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider: Add crypto 'Croatian eID' 4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f","level":"info"}
{"message":"Provider:AddCrypto: PKCS#11 'C:\\Program Files\\AKD\\eID Middleware\\pkcs11\\AkdEidPkcs11_64.dll' 'Croatian eID'","level":"info"}
{"message":"Provider:Token Amount of tokens was changed (+2/-0)","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/login","level":"info"}
{"message":"Fortify: Create window p11-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
{"message":"Server: Close session  origin:https://tools.fortifyapp.com (code: 1001)","level":"info"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:false","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/login","level":"info"}
{"message":"Fortify: Create window key-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/login","level":"info"}
{"message":"Fortify: Create window p11-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/export","level":"info"}
{"message":"Server: Close session  origin:https://tools.fortifyapp.com (code: 1001)","level":"info"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:false","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/login","level":"info"}
{"message":"Fortify: Create window key-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
{"message":"Server: Close session  origin:https://tools.fortifyapp.com (code: 1001)","level":"info"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:false","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/login","level":"info"}
{"message":"Fortify: Create window key-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/certificateStorage/export","level":"info"}
{"message":"TypeError: Cannot assign to read only property 'extractable' of object '#<CryptoKey>'\n    at FileStorage.ecKeyToBase64 (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:233:29)\n    at FileStorage.saveIdentities (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:165:61)\n    at processTicksAndRejections (internal/process/task_queues.js:86:5)\n    at async FileStorage.getIdentity (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:193:13)\n    at async Server.getRandomBundle (C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:604:26)\n    at async C:\\Program Files\\Fortify\\resources\\app.asar\\node_modules\\@webcrypto-local\\server\\build\\index.js:445:40","level":"error"}
{"message":"Server: Close session  origin:https://tools.fortifyapp.com (code: 1001)","level":"info"}
{"message":"Server: New session connect https://tools.fortifyapp.com","level":"info"}
{"message":"Server: Push session to stack","level":"info"}
{"message":"Server: Cannot parse MessageSignedProtocol","level":"info"}
{"message":"Server: Initialize secure session origin:https://tools.fortifyapp.com id:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d authorized:false","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d server/login","level":"info"}
{"message":"Fortify: Create window key-pin","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/info","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a27fd2c5a84c4e4904bf145d2bf4e2d3fdd7e6b6 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:a7370eae6951997646c5bfedf8f3df0d8b8b698d crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:0a86f07e3fb6325839dfcd1d8a5e4ddfd6454a7d202b6cda30c6045ee14324c4 crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider/action/getCrypto","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/isLoggedIn","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/keyStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/keys","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/getItem","level":"info"}
{"message":"Server: session:b4ac4ce6c005e5d903800c99aead9e82fc78458476129dace693b2c9da3d6a0d provider:4458a3c0a29183e4eb00a92d23036a2670d7b874998f15ae529a7f0e00a9694f crypto/certificateStorage/export","level":"info"}
```

**Comment by microshine** [comment]
@StipT I found `Cannot assign to read only property 'extractable' of object '#<CryptoKey>'` error today. It occurs of `@peculiar/webcrypto` secure update. We downgraded that module to prev version
https://drive.google.com/open?id=18psGwBHO7FtH8HafFvO5Hy98auMidc4c

**Comment by StipT** [comment]
@microshine Do you have .msi version?

**Comment by microshine** [comment]
Sorry. Forgot about it
Give me 15 minutes to build MSI

**Comment by microshine** [comment]
@StipT We updated Electron to the latest version. It'll take more time to update Windows build.

**Comment by microshine** [comment]
https://drive.google.com/open?id=1a2Sp4EQp1Ae8J0PHCAW-G-dMR1XZkeo0

**Comment by StipT** [comment]
It works great now. 
Thank you for the awesome support!

**Comment by StipT** [comment]
You can add this, it works great.
```
	{
   	"atr": "3BFF1300008131FE450031B9640444ECC17394018082900012",
   	"name": "Croatian eID",
   	"driver": "8CAFDB28B1B85D83853A40635B9AFBD2BB18F9E6"
   }
```
```
	{
   	"id": "8CAFDB28B1B85D83853A40635B9AFBD2BB18F9E6",
   	"name": "Croatian eID",
   	"file": {
   		"windows": "%PROGRAMFILES/AKD/eID Middleware/pkcs11/AkdEidPkcs11_64.dll",
		"linux": "/usr/lib/akd/eidmiddleware/pkcs11/libEidPkcs11.so",
   		"osx": "/usr/local/lib/pkcs11/libEidPkcs11.dylib"	
   	}
   }
```

**Comment by microshine** [comment]
@StipT I'll prepare release version and publish it. Right now you've got a demo version

**Comment by StipT** [comment]
@microshine has it been added yet?
